go/logic/topology_recovery.go

```diff
@@ -1479,14 +1479,124 @@ func checkAndRecoverNonWriteableMaster(analysisEntry inst.ReplicationAnalysis, c
 
 // checkAndRecoverLockedSemiSyncMaster
 func checkAndRecoverLockedSemiSyncMaster(analysisEntry inst.ReplicationAnalysis, candidateInstanceKey *inst.InstanceKey, forceInstanceRecovery bool, skipProcesses bool) (recoveryAttempted bool, topologyRecovery *TopologyRecovery, err error) {
-
 	topologyRecovery, err = AttemptRecoveryRegistration(&analysisEntry, true, true)
 	if topologyRecovery == nil {
 		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("found an active or recent recovery on %+v. Will not issue another RecoverLockedSemiSyncMaster.", analysisEntry.AnalyzedInstanceKey))
 		return false, nil, err
 	}
 
-	return false, nil, nil
+	if !config.Config.EnforceSemiSyncReplicaCount {
+		return false, nil, nil
+	}
+
+	replicas, err := inst.ReadReplicaInstances(&analysisEntry.AnalyzedInstanceKey)
+	if err != nil {
+		return false, topologyRecovery, err
+	}
+	if len(replicas) == 0 {
+		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("no replicas found for %+v.", analysisEntry.AnalyzedInstanceKey))
+		return false, topologyRecovery, nil
+	}
+
+	reversePriorityReplicas := replicas // TODO sort replicas by priority
+	enableCount := int(analysisEntry.SemiSyncMasterWaitForReplicaCount) - int(analysisEntry.CountSemiSyncReplicasEnabled)
+	if enableCount <= 0 || enableCount > len(replicas) {
+		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("invalid state for recovery %+v. Wait count is lower or equal to semi sync replica count or replica count too low.", analysisEntry.AnalyzedInstanceKey))
+		return false, nil, err
+	}
+
+	// No way back. We're doing this.
+
+	// TODO before actually doing it, do sanity check (downtimed, etc.)
+
+	enabled := 0
+	for _, replica := range reversePriorityReplicas {
+		if replica.IsDowntimed || replica.SemiSyncReplicaEnabled {
+			continue
```

Comment:
We might actually want to _disable_ semi-sync when we have more than `enough` semi sync replicas enables.

Reply:
To clarify the previous comment: we might have the exact number of desired semi-sync replicas, but wish to reassign to different replicas. As we assign semi-sync to a replica, we may wish to deassign the role from a lower-priority replica.

Reply:
I have implemented this in the latest PR (still WIP). It now orders the replicas (see https://github.com/openark/orchestrator/pull/1373#discussion_r660566882) and turns semi-sync on/off based on the desired state. 

Aside from `LockedSemiSyncMaster` (too few) and `MasterWithTooManySemiSyncReplicas` (too many), there is also the situation of "the wrong ones" now that may need its own analysis code, maybe `MasterWithIncorrectSemiSyncReplicas`. I'll add that to the TODO list. The remediation for all of them is the same, i.e. the logic in `recoverExactSemiSyncReplicas()`.